funsor/torch/provenance.py

```diff
@@ -0,0 +1,65 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+
+
+class ProvenanceTensor(torch.Tensor):
+    """
+    Provenance tracking implementation in Pytorch.
+
+    Provenance of the output tensor is the union of provenances of input tensors.
+    """
+
+    def __new__(cls, data, provenance=frozenset(), **kwargs):
+        if not provenance:
+            return data
+        instance = torch.Tensor.__new__(cls)
+        instance.__init__(data, provenance)
+        return instance
+
+    def __init__(self, data, provenance=frozenset()):
+        assert isinstance(provenance, frozenset)
+        if isinstance(data, ProvenanceTensor):
+            provenance |= data._provenance
+            data = data._t
+        self._t = data
+        self._provenance = provenance
+
+    def __repr__(self):
+        return "Provenance:\n{}\nTensor:\n{}".format(self._provenance, self._t)
+
+    def __torch_function__(self, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        # collect provenance information from args
+        provenance = frozenset()
+        # extract ProvenanceTensor._t data from args
+        _args = []
+        for arg in args:
```

Comment:
This logic is a bit convoluted. Maybe it could be simplified with some of the helpers in [`torch.overrides`](https://pytorch.org/docs/stable/torch.overrides.html)?

Reply:
> the helpers in torch.overrides?

That might be  useful, I look more into `torch.overrides` functionality.